app/src/common/shared/com/igalia/wolvic/utils/DictionariesManager.java

```diff
@@ -0,0 +1,153 @@
+package com.igalia.wolvic.utils;
+
+import android.content.Context;
+import android.os.Build;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.igalia.wolvic.downloads.Download;
+import com.igalia.wolvic.downloads.DownloadJob;
+import com.igalia.wolvic.downloads.DownloadsManager;
+import com.igalia.wolvic.ui.widgets.WidgetManagerDelegate;
+
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.Objects;
+
+public class DictionariesManager implements DownloadsManager.DownloadsListener {
+
+    static final String LOGTAG = SystemUtils.createLogtag(DictionariesManager.class);
+
+    private final Context mContext;
+    private final DownloadsManager mDownloadManager;
+    private long mDicDownloadLang = -1;
+
+    public DictionariesManager(@NonNull Context context) {
+        mContext = context;
+        WidgetManagerDelegate mApplicationDelegate = ((WidgetManagerDelegate) context);
+        mDownloadManager = mApplicationDelegate.getServicesProvider().getDownloadsManager();
+    }
+
+    public void init() {
+        mDownloadManager.addListener(this);
+    }
+
+    public void end() {
+        mDownloadManager.removeListener(this);
+    }
+
+    @Nullable
+    public String getOrDownloadDictionary(@NonNull String lang) {
+        String dictionaryPath = null;
+        if (DictionaryUtils.isBuiltinDictionary(lang)) {
+            dictionaryPath = DictionaryUtils.getBuiltinDicPath();
+
+            for (String dbName : DictionaryUtils.getBuiltinDicNames(lang)) {
+                if (!mContext.getDatabasePath(dbName).exists()) {
+                    try {
+                        InputStream in = mContext.getAssets().open(dictionaryPath + dbName);
+                        storeDatabase(in, dbName);
+                    } catch (Exception e) {
+                        Log.e(LOGTAG, Objects.requireNonNull(e.getMessage()));
+                    }
+                }
+            }
+        } else if (DictionaryUtils.isExternalDictionary(mContext, lang)) {
+            dictionaryPath = DictionaryUtils.getExternalDicPath(mContext, lang);
+            if (dictionaryPath != null) return dictionaryPath;
+
+            if (mDicDownloadLang != -1) {
+                mDownloadManager.removeDownload(mDicDownloadLang, true);
+            }
+
+            downloadDictionary(lang);
+        }
+
+        return dictionaryPath;
+    }
+
+    private void downloadDictionary(@NonNull String lang) {
+        final Dictionary dictionary = DictionaryUtils.getExternalDictionaryByLang(mContext, lang);
+        final String payload = DictionaryUtils.getDictionaryPayload(dictionary);
```

Comment:
you're using dictionary here without any check. It can be null

Reply:
Fixed and opened the PR #1175 to fix the same thing in Environments manager.